geo/src/algorithm/extremes.rs

```diff
@@ -1,274 +1,131 @@
+/// assert_eq!(extremes.y_max.coord.y, 2.);
+/// ```
+pub trait Extremes<'a, T: CoordinateType> {
+    fn extremes(&'a self) -> Option<Outcome<T>>;
 }
 
-// wrapper for extreme-finding function
-fn find_extreme_indices<T, F>(func: F, polygon: &Polygon<T>) -> Result<Extremes, ()>
-where
-    T: HasKernel + Signed,
-    F: Fn(Coordinate<T>, &Polygon<T>) -> Result<usize, ()>,
-{
-    if !polygon.exterior().is_convex() {
-        return Err(());
-    }
-    let directions: Vec<Coordinate<_>> = vec![
-        (T::zero(), -T::one()).into(),
-        (T::one(), T::zero()).into(),
-        (T::zero(), T::one()).into(),
-        (-T::one(), T::zero()).into(),
-    ];
-    Ok(directions
-        .into_iter()
-        .map(|p| func(p, polygon).unwrap())
-        .collect::<Vec<usize>>()
-        .into())
-}
-
-// find a convex, counter-clockwise oriented polygon's maximum vertex in a specified direction
-// u: a direction vector. We're using a point to represent this, which is a hack but works fine
-fn polymax_naive_indices<T>(u: Coordinate<T>, poly: &Polygon<T>) -> Result<usize, ()>
-where
-    T: HasKernel + Signed,
-{
-    let vertices = &poly.exterior().0;
-    let mut max: usize = 0;
-    for (i, _) in vertices.iter().enumerate() {
-        // if vertices[i] is above prior vertices[max]
-        if above(u, vertices[i], vertices[max]) {
-            max = i;
-        }
-    }
-    Ok(max)
+#[derive(Debug, PartialEq)]
+pub struct Extreme<T: CoordinateType> {
+    pub index: usize,
+    pub coord: Coordinate<T>,
 }
 
-pub trait ExtremeIndices {
-    /// Find the extreme `x` and `y` _indices_ of a convex Polygon
-    ///
-    /// The polygon **must be convex and properly (ccw) oriented**.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use geo::extremes::ExtremeIndices;
-    /// use geo::polygon;
-    ///
-    /// // a diamond shape
-    /// let polygon = polygon![
-    ///     (x: 1.0, y: 0.0),
-    ///     (x: 2.0, y: 1.0),
-    ///     (x: 1.0, y: 2.0),
-    ///     (x: 0.0, y: 1.0),
-    ///     (x: 1.0, y: 0.0),
-    /// ];
-    ///
-    /// // Polygon is both convex and oriented counter-clockwise
-    /// let extremes = polygon.extreme_indices().unwrap();
-    ///
-    /// assert_eq!(extremes.ymin, 0);
-    /// assert_eq!(extremes.xmax, 1);
-    /// assert_eq!(extremes.ymax, 2);
-    /// assert_eq!(extremes.xmin, 3);
-    /// ```
-    fn extreme_indices(&self) -> Result<Extremes, ()>;
+#[derive(Debug, PartialEq)]
+pub struct Outcome<T: CoordinateType> {
+    pub x_min: Extreme<T>,
+    pub y_min: Extreme<T>,
+    pub x_max: Extreme<T>,
+    pub y_max: Extreme<T>,
 }
 
-impl<T> ExtremeIndices for Polygon<T>
+impl<'a, T, G> Extremes<'a, T> for G
 where
-    T: Signed + HasKernel,
+    G: CoordsIter<'a, Scalar = T>,
+    T: CoordinateType,
 {
-    fn extreme_indices(&self) -> Result<Extremes, ()> {
-        find_extreme_indices(polymax_naive_indices, self)
-    }
-}
+    fn extremes(&'a self) -> Option<Outcome<T>> {
```

Comment:
I like the new API! 

How arduous would it be to maintain a deprecated version of the old API for compatibility? 

I know we're pre-v1 so technically we can "do whatever we want", but whenever we bump semver (for pre v1 like us, that's the `X` in `0.X.Y`), people get left behind for a while, which means it's more likely they're using old code until their crate maintainers update, or are stuck with multiple incompatible versions.

That's why when it's easy, I'd prefer to do things in a way that uses deprecation so downstream users don't get stranded or need multiple versions of geo for compatibility reasons. Then we can batch all the breaking changes together, by deleting the deprecated methods, only once in a while.

I also think in-line deprecation notices are more helpful upgrade documentation than release notes for the people who update before the deprecated methods are removed.

(If it's not easy, then I'm OK with merging as is and bumping the minor version)

Reply:
Before I started implementing this, I took a glance at these GitHub queries: [1](https://github.com/search?l=rust&q=geo+extremepoints&type=code) [2](https://github.com/search?l=rust&q=geo+extremeindices&type=code). Assuming these are doing the right thing, there's (essentially) no open source project utilizing the old traits. There could be users, but we at least have some indication there's minimal breakage in the GeoRust ecosystem. And I also took a look at the [current changelog](https://github.com/georust/geo/blob/master/geo/CHANGES.md) which already has at least [one breaking change](https://github.com/georust/geo/pull/593), so no matter what we're gonna need to bump to 0.17.0 with the next release (unless we were to branch off master). So I'd say it's not difficult to do the deprecation dance, and I do understand the benefits, but it still takes time for myself and other maintainers. So all in all I'm neutral about doing it in this case, but I have capacity to do it if you still prefer.